core-bundle/src/Filesystem/FilesystemItemIterator.php

```diff
@@ -0,0 +1,83 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Filesystem;
+
+/**
+ * @implements \IteratorAggregate<int, FilesystemItem>
+ */
+class FilesystemItemIterator implements \IteratorAggregate
+{
+    /**
+     * @var iterable<FilesystemItem>
+     */
+    private iterable $listing;
+
+    /**
+     * @param iterable<FilesystemItem> $listing
+     */
+    public function __construct(iterable $listing)
+    {
+        $this->listing = $listing;
+    }
+
+    /**
+     * @param callable(FilesystemItem):bool $filter
+     */
+    public function filter(callable $filter): self
+    {
+        $listFiltered = static function (iterable $listing) use ($filter): \Generator {
+            foreach ($listing as $item) {
+                if ($filter($item)) {
+                    yield $item;
+                }
+            }
+        };
+
+        return new self($listFiltered($this->listing));
+    }
+
+    public function files(): self
+    {
+        return $this->filter(static fn (FilesystemItem $item) => $item->isFile());
+    }
+
+    public function directories(): self
+    {
+        return $this->filter(static fn (FilesystemItem $item) => !$item->isFile());
+    }
+
+    /**
+     * @return \Traversable<FilesystemItem>
+     */
+    public function getIterator(): \Traversable
+    {
+        foreach ($this->listing as $item) {
+            if (!$item instanceof FilesystemItem) {
+                /** @phpstan-ignore-next-line */
+                $type = \is_object($item) ? \get_class($item) : \gettype($item);
```

Comment:
You cannot use `get_class` on literals and using `gettype` on objects only results in "object". So this way you'll always get a meaningful message.

Reply:
So why did you have to add `@phpstan-ignore-next-line`?

Reply:
Because trusting the annotated types, this case would never happen. But it's a long way until we'll eventually get real generics, so I followed @ausi's point in https://github.com/contao/contao/pull/4012#discussion_r793088830 and added a type check.

Reply:
failing run before ignoring: https://github.com/contao/contao/runs/4964774085?check_suite_focus=true